patches/server/1065-API-to-allow-disallow-tick-sleeping.patch

```diff
@@ -0,0 +1,74 @@
+index 64b56abf8900d0424100da460fc68ac964394793..adaac226886b5ceec3b6228904d784e9773c4536 100644
+--- a/src/main/java/net/minecraft/server/MinecraftServer.java
++++ b/src/main/java/net/minecraft/server/MinecraftServer.java
+@@ -307,6 +307,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
+     private static final AtomicReference<RuntimeException> fatalException = new AtomicReference();
+     private final SuppressedExceptionCollector suppressedExceptions;
+     private final DiscontinuousFrame tickFrame;
++    private boolean allowPausing = true; // Paper - API to allow/disallow tick sleeping
+ 
+     // CraftBukkit start
+     public final WorldLoader.DataLoadContext worldLoader;
+@@ -1624,7 +1625,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
+         int j = this.pauseWhileEmptySeconds() * 20;
+ 
+         if (j > 0) {
+-            if (this.playerList.getPlayerCount() == 0 && !this.tickRateManager.isSprinting()) {
++            if (this.playerList.getPlayerCount() == 0 && !this.tickRateManager.isSprinting() && allowPausing) { // Paper - API to allow/disallow tick sleeping
+                 ++this.emptyTicks;
+             } else {
+                 this.emptyTicks = 0;
+@@ -3186,4 +3187,11 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
+         }
+     }
+     // Paper end - Add tick times API and /mspt command
++
++    // Paper start - API to allow/disallow tick sleeping
++    public void allowPausing(boolean value) {
++        allowPausing = value;
++    }
++    // Paper end - API to allow/disallow tick sleeping
++
+ }
+diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+index dfddcfb1fe1679adaecf75375757dca720e76ce1..6f2a2e63932554da2296f7aa050b832a00051438 100644
+--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
++++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+@@ -258,6 +258,7 @@ import org.bukkit.scoreboard.Criteria;
+ import org.bukkit.structure.StructureManager;
+ import org.bukkit.util.StringUtil;
+ import org.bukkit.util.permissions.DefaultPermissions;
++import org.jetbrains.annotations.NotNull;
+ import org.yaml.snakeyaml.LoaderOptions;
+ import org.yaml.snakeyaml.Yaml;
+ import org.yaml.snakeyaml.constructor.SafeConstructor;
+@@ -313,6 +314,7 @@ public final class CraftServer implements Server {
+     private final io.papermc.paper.logging.SysoutCatcher sysoutCatcher = new io.papermc.paper.logging.SysoutCatcher(); // Paper
+     private final io.papermc.paper.potion.PaperPotionBrewer potionBrewer; // Paper - Custom Potion Mixes
+     public final io.papermc.paper.SparksFly spark; // Paper - spark
++    private List<org.bukkit.plugin.java.JavaPlugin> pluginsBlockingSleep = new ArrayList<>(); // Paper - API to allow/disallow tick sleeping
```

Comment:
Storing plugin instances makes memory leaks easy, it might be better to store their names instead and remove entries during plugin disables/unloads. You can also just move the list to MinecraftServer and check if it's empty there instead of adding the extra boolean to MinecraftServer

Reply:
Will do

Reply:
But should I leave the instance part in the API? I think this is more up to standard with the API as a whole and simpler.

Reply:
Yup, only talking about the implementation here

Reply:
I haven't made very many commits yet, so I don't really know exactly how every part works. Do you maybe know a good place to add a call, as there are around 20 onDisable related functions and events

Reply:
I think it should be fine to just iterate the list if it's not empty in tickServer, removing any entries that point to missing or disabled plugins

Reply:
Wouldn't that be less efficient? Checking every tick if every plugin in the list is enabled sounds not so efficient to me

Reply:
not enough to make it relevant

1) it's just a map lookup and field read
2) we only iterate when it's non-empty, so during tick sleep and rarely when an invalid entry gets removed

Reply:
Can probably just use removeIf instead of iterating manually, should keep the isEmpty check though

Reply:
I have made the change with the removeIf() implementation

Reply:
Will now also rebase